### PR TITLE
xcodebuild output handling

### PR DESCRIPTION
**Pull request by memega**
- redesigned handling of remote command output, introducing timeout interruption;
- improved coroutines support with accurate structure;
- after a test run failure, simulator provider attempts to reuse the same simulator
- updated the sample app;
- added custom log configuraton;
- added a shutdown hook to print timeline summary even when execution is interrupted.

### CODE REVIEW

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileConfiguration.kt [hunk 1]
```diff
@@ -31,6 +31,7 @@ data class FileConfiguration(
         var includeSerialRegexes: Collection<Regex>?,
         var excludeSerialRegexes: Collection<Regex>?,
 
+        var timeoutMillis: Long?,
```
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileConfiguration.kt]: Can we create a more meaningful name? I suspect this is the whole run timeout

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt [hunk 2]
```diff
@@ -23,7 +23,12 @@ data class FileIOSConfiguration(
         @JsonProperty("knownHostsPath") val knownHostsPath: File?,
         @JsonProperty("remoteRsyncPath") val remoteRsyncPath: String = "/usr/bin/rsync",
         @JsonProperty("sourceRoot") val sourceRoot: File?,
+        @JsonProperty("alwaysEraseSimulators") val alwaysEraseSimulators: Boolean?,
         @JsonProperty("debugSsh") val debugSsh: Boolean?,
+        @JsonProperty("hideRunnerOutput") val hideRunnerOutput: Boolean?,
+        @JsonProperty("compactOutput") val compactOutput: Boolean = false,
+        @JsonProperty("teamcityCheckoutDir") val teamcityCheckoutDir: File? = null,
```
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt]: We should not depend on the TeamCity in the marathon. And marathon can also be executed locally. Let's rename this as well

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/MarathonCliConfiguration.kt [hunk 3]
```diff
@@ -8,9 +8,6 @@ class MarathonCliConfiguration(parser: ArgParser) {
     val marathonfile: File by parser
             .storing("--marathonfile", "-m", help = "marathonfile file path") { File(this) }
             .default(File("Marathonfile"))
-    val androidSdkDir: File? by parser
```
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/args/MarathonCliConfiguration.kt:11]: :+1:

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt [hunk 4]
```diff
@@ -20,6 +25,7 @@ internal class InfluxDbTracker(private val influxDb: InfluxDB) : NoOpTracker() {
                 .tag("class", testResult.test.clazz)
                 .tag("method", testResult.test.method)
                 .tag("deviceSerial", device.serialNumber)
+                .tag("run", InfluxDbTracker.runUUID)
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt]: This is high cardinality tag. I'm not sure we want to make it a tag.

File: core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceAction.kt [hunk 5]
```diff
@@ -8,5 +8,5 @@ sealed class DeviceAction {
     object Initialize : DeviceAction()
     data class Terminate(val batch: TestBatch? = null) : DeviceAction()
     data class ExecuteBatch(val batch: TestBatch, val result: CompletableDeferred<TestBatchResults>) : DeviceAction()
-    data class RequestNextBatch(val result: CompletableDeferred<TestBatchResults>? = null) : DeviceAction()
+    data class NotifyIsReady(val result: CompletableDeferred<TestBatchResults>? = null) : DeviceAction()
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceAction.kt:11]: :+1:

File: core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt [hunk 6]
```diff
@@ -74,8 +74,24 @@ class DevicePoolActor(private val poolId: DevicePoolId,
...
     }
 
-    private suspend fun deviceReady(msg: DevicePoolMessage.FromDevice.RequestNextBatch) {
-        queue.send(QueueMessage.RequestBatch(msg.device))
+    private suspend fun deviceReady(msg: DevicePoolMessage.FromDevice.IsReady) {
+        maybeRequestBatch(msg.device)
+    }
+
+    private suspend fun maybeRequestBatch(avoiding: Device? = null) {
+        val shuffledDevices = devices.values
```
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/execution/DevicePoolActor.kt]: Can we use sequences? https://blog.kotlin-academy.com/effective-kotlin-use-sequence-for-bigger-collections-with-more-than-one-processing-step-649a15bb4bf

File: vendor/vendor-ios/src/main/kotlin/com/malinskiy/marathon/ios/device/LocalListSimulatorProvider.kt [hunk 7]
```diff
@@ -3,65 +3,141 @@ package com.malinskiy.marathon.ios.device
...
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.launch
-import kotlinx.coroutines.newFixedThreadPoolContext
-import java.io.File
-import java.net.InetAddress
+import kotlinx.coroutines.withContext
+import java.util.concurrent.ConcurrentHashMap
 import kotlin.coroutines.CoroutineContext
 
+private const val MAX_SERIAL = 8
```
Comment (Malinskiy) [inline vendor/vendor-ios/src/main/kotlin/com/malinskiy/marathon/ios/device/LocalListSimulatorProvider.kt]: The purpose is unclear. Please rename

**Review by Malinskiy — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt (hunk 4 above)
Comment (tagantroy) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt]: > fields are not indexed. tags are indexed.

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt (hunk 2 above)
Comment (memega) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt]: First of all, this is an optional value; secondly, it is used _only_ to produce teamcity service messages

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileConfiguration.kt (hunk 1 above)
Comment (memega) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileConfiguration.kt]: That would be a single command execution timeout

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt (hunk 4 above)
Comment (memega) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/tracker/remote/influx/InfluxDbTracker.kt]: Removed it altogether

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt (hunk 2 above)
Comment (memega) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/args/FileIOSConfiguration.kt]: Removed it along with unfinished logarchive collecting feature

**Pull request merged**